### PR TITLE
kubeconfig.write: update the issuer with ingressPort

### PR DESCRIPTION
**Pull request by vyasgun**
**Fixes:** Issue #4124 

**Relates to:** Issue #4177 (_Maybe_)

## Solution/Idea

This is a workaround to explicitly set the `ingressPortNumber` in the `Issuer` field when requesting token. By default, the OpenShift library uses the URL from `/.well-known/oauth-authorization-server` without including the custom port set by the user on CRC. This omission causes errors when writing new contexts to kubeconfig, as the token request fails due to the missing port number. Since users configure this custom port for their CRC instance, this commit ensures the issuer field includes the ingress-https-port before making the token request.
## Proposed changes

List main as well as consequential changes you introduced or had to introduce.

1. Wrote a new equivalent local function to perform the port setting.

## Testing

1. `crc config set ingress-https-port 7443`
2. `crc setup`
3. `crc start`

### CODE REVIEW

**Comment by praveenkumar** [comment]
Can you put more info in the commit logs why it was not working without this PR and what this PR adds to make it work?

**Comment by vyasgun** [comment]
Thanks @praveenkumar. Just added more info to the commit log.

File: pkg/crc/machine/kubeconfig.go [hunk 1]
```diff
@@ -198,13 +198,23 @@ func getTokenForUser(username, password, ip string, ca []byte, clusterConfig *ty
...
 	return token, nil
 }
 
+func requestTokenWithChallengeHandlers(clientCfg *restclient.Config, handler *challengehandlers.BasicChallengeHandler, port uint) (string, error) {
+	o, err := tokenrequest.NewRequestTokenOptions(clientCfg, false).WithChallengeHandlers(handler)
+	if err != nil {
+		return "", err
+	}
+
+	o.Issuer = fmt.Sprintf("%s:%d", o.Issuer, port)
```
Comment (praveenkumar) [inline pkg/crc/machine/kubeconfig.go]: if we are adding the port explicit to issuer then I don't think we really need https://github.com/crc-org/crc/pull/4183/files#diff-a796f5c0a3b2fdd2f85be66107ff80578a3658029bd675c39ac6c015b515ebecR188-R190 part so let's try to remove that and see if that works.

File: pkg/crc/machine/kubeconfig.go [hunk 2]
```diff
@@ -181,13 +180,10 @@ func getTokenForUser(username, password, ip string, ca []byte, clusterConfig *ty
...
 			DialContext: func(ctx gocontext.Context, network, address string) (net.Conn, error) {
 				logging.Debugf("Using address: %s", address)
-				hostname, port, err := net.SplitHostPort(address)
+				_, port, err := net.SplitHostPort(address)
 				if err != nil {
 					return nil, err
 				}
-				if strings.HasSuffix(hostname, constants.AppsDomain) {
-					port = strconv.Itoa(int(ingressHTTPSPort))
-				}
```
Comment (praveenkumar) [inline pkg/crc/machine/kubeconfig.go]: @vyasgun After removing it looks like it fails again, did you test this scenario?
```
DEBU Using address: api.crc.testing:6443          
DEBU Dialing to 127.0.0.1:6443                    
DEBU Using address: oauth-openshift.apps-crc.testing:443 
DEBU Dialing to 127.0.0.1:443                     
ERRO Cannot update kubeconfig: dial tcp 127.0.0.1:443: connect: connection refused 
DEBU Making call to close driver server           
```

Comment (vyasgun) [reply pkg/crc/machine/kubeconfig.go]: I didn't face the issue. I will take another look and update.

Comment (vyasgun) [reply pkg/crc/machine/kubeconfig.go]: I added the lines back and updated the PR.

File: pkg/crc/machine/kubeconfig.go [hunk 3]
```diff
@@ -10,7 +10,7 @@ import (
 	"net/url"
 	"os"
 	"path/filepath"
-	"strconv"
+	`strconv`
```
Comment (anjannath) [inline pkg/crc/machine/kubeconfig.go]: unrelated change probably by the ide?

Comment (praveenkumar) [reply pkg/crc/machine/kubeconfig.go]: @vyasgun can you fix it and repush it?

Comment (gbraad) [reply pkg/crc/machine/kubeconfig.go]: @vyasgun can you handle this ASAP, otherwise it will not be possible to merge in time for the next release

Comment (gbraad) [reply pkg/crc/machine/kubeconfig.go]: did linting fail? otherwise we can merge regardless and fix this ASAP if not done today.

Comment (cfergeau) [reply pkg/crc/machine/kubeconfig.go]: This PR has " Maintainers are allowed to edit this pull request.", I've fixed this.

File: pkg/crc/machine/kubeconfig.go (hunk 1 above)
Comment (cfergeau) [inline pkg/crc/machine/kubeconfig.go]: Using https://pkg.go.dev/net#JoinHostPort may be slightly more correct. But there is this issue throughout the codebase, all of these can be changed in one go the day we look at proper ipv6 support.

Comment (vyasgun) [reply pkg/crc/machine/kubeconfig.go]: thanks, will change this.

File: pkg/crc/machine/kubeconfig.go (hunk 3 above)
Comment (vyasgun) [reply pkg/crc/machine/kubeconfig.go]: thanks @cfergeau! Sorry I was AFK

**Review by gbraad — APPROVED** [review]

File: pkg/crc/machine/kubeconfig.go (hunk 1 above)
Comment (praveenkumar) [reply pkg/crc/machine/kubeconfig.go]: This was mistake because `o.Issuer` is URL and not the host string and because of that if take it as ip6 and fails
```
ERRO Cannot update kubeconfig: parse "[https://oauth-openshift.apps-crc.testing]:443": first path segment in URL cannot contain colon 
```
